ui/src/spinbox.tsx

```diff
@@ -96,35 +96,37 @@ export const
             : value < min
               ? min
               : value
-        wave.args[name] = newValue
-        if (trigger) wave.push()
-        if (precision > 0 && isLastCharDotOrTraillingZero.test(val)) {
+        if (val === '-') setVal('-')
+        else if (!precision) setVal(String(newValue).split('.')[0])
+        else if (isLastCharDotOrTraillingZero.test(val)) {
           // We can't use parseValue because it requires casting to number which will remove the trailling zeros.
           const [head, tail = ''] = val.split('.')
           setVal(`${head}.${tail.slice(0, precision)}`)
-        } else {
-          setVal(String(newValue))
         }
-      }, [max, min, name, parseValue, precision, trigger]),
+        else setVal(String(newValue))
+        return newValue
+      }, [max, min, parseValue, precision]),
+      handleOnInput = React.useCallback((val: U) => {
+        wave.args[name] = val
+        if (trigger) wave.push()
+      }, [name, trigger]),
       debouncedHandleOnInput = React.useRef(wave.debounce(DEBOUNCE_TIMEOUT, handleOnInput)),
       onInput = (e: React.SyntheticEvent<HTMLElement>) => {
-        const val = (e.target as HTMLInputElement).value
-        setVal(val)
-        trigger ? debouncedHandleOnInput.current(val) : handleOnInput(val)
+        const numVal = handleValue((e.target as HTMLInputElement).value)
+        trigger ? debouncedHandleOnInput.current(numVal) : handleOnInput(numVal)
       }
 
     // eslint-disable-next-line react-hooks/exhaustive-deps
     React.useEffect(() => { wave.args[name] = (value < min) ? min : ((value > max) ? max : value) }, [])
 
     return (
       <Fluent.SpinButton
-        inputProps={{ 'data-test': name } as React.InputHTMLAttributes<HTMLInputElement>}
+        inputProps={{ 'data-test': name, value: val } as React.InputHTMLAttributes<HTMLInputElement>}
```

Comment:
Are you passing `value` directly to the native input element to prevent SpinButton from controlling `value`?

Reply:
Yes